nexus/src/app/background/tasks/tuf_artifact_replication.rs

```diff
@@ -0,0 +1,447 @@
+                    // If the last sender for this channel is dropped, then the
+                    // `Nexus` type has also been dropped. This is presumably
+                    // a bug.
+                    panic!("artifact replication receiver disconnected");
+                }
+            }
+        }
+
+        // Query the database for artifacts that we ought to have.
+        let mut artifacts = HashMap::new();
+        let mut paginator = Paginator::new(SQL_BATCH_SIZE);
+        while let Some(p) = paginator.next() {
+            let batch = datastore
+                .update_tuf_artifact_list(opctx, &p.current_pagparams())
+                .await?;
+            paginator = p.found_batch(&batch, &|a| a.sha256);
+            for artifact in batch {
+                artifacts
+                    .insert(artifact.sha256.0, ArtifactPresence::default());
+            }
+        }
+        // Ask all sled agents to list the artifacts they have, and mark those
+        // artifacts as present on those sleds.
+        let sleds = datastore
+            .sled_list_all_batched(&opctx, SledFilter::InService)
+            .await?
+            .into_iter()
+            .map(|sled| Sled {
+                id: SledUuid::from_untyped_uuid(sled.identity.id),
+                client: sled_client_from_address(
+                    sled.identity.id,
+                    sled.address(),
+                    log,
+                ),
+                depot_base_url: format!(
+                    "http://{}",
+                    sled.address_with_port(sled.repo_depot_port.into())
+                ),
+            })
+            .collect::<Vec<_>>();
+        ensure!(!sleds.is_empty(), "no sleds");
+        let responses =
+            futures::future::join_all(sleds.iter().map(|sled| async move {
+                let response = match sled.client.artifact_list().await {
+                    Ok(response) => response.into_inner(),
+                    Err(err) => {
+                        error!(
+                            log,
+                            "Failed to get artifact list";
+                            "error" => InlineErrorChain::new(&err),
+                            "sled" => sled.client.baseurl(),
+                        );
+                        HashMap::new()
+                    }
+                };
+                (sled, response)
+            }))
+            .await;
+        let mut delete_requests = Vec::new();
+        for (sled, response) in responses {
+            for (hash, count) in response {
+                if let Some(presence) = ArtifactHash::from_str(&hash)
+                    .ok()
+                    .and_then(|hash| artifacts.get_mut(&hash))
+                {
+                    presence.counts.insert(sled.id, count);
+                    presence.sleds.push(sled);
+                } else {
+                    delete_requests.push(Request::Delete { sled, hash });
```

Comment:
We will already emit an `info!` or higher-severity message if this request occurs (at the bottom of the `execute` function). I think this would unnecessarily spam the logs if there's any artifacts for deletion.